tests/EventTest.php

```diff
@@ -119,32 +119,27 @@ public function testToArray(): void
         $this->assertEquals($expected, $event->toArray());
     }
 
-    public function testToArrayWithMessage(): void
+    /**
+     * @dataProvider toArrayWithMessageDataProvider
+     */
+    public function testToArrayWithMessage(array $message, $expectedValue): void
     {
         $event = new Event();
-        $event->setMessage('foo bar');
+        call_user_func_array([$event, 'setMessage'], $message);
 
         $data = $event->toArray();
 
         $this->assertArrayHasKey('message', $data);
-        $this->assertEquals('foo bar', $data['message']);
+        $this->assertSame($expectedValue, $data['message']);
     }
 
-    public function testToArrayWithMessageWithParams(): void
+    public function toArrayWithMessageDataProvider(): array
     {
-        $expected = [
-            'message' => 'foo %s',
-            'params' => ['bar'],
-            'formatted' => 'foo bar',
+        return [
+            [['foo bar'], 'foo bar'],
+            [['foo %s', ['bar']], ['message' => 'foo %s', 'params' => ['bar'], 'formatted' => 'foo bar']],
+            [['foo @bar', ['@bar' => 'bar'], 'foo bar'], ['message' => 'foo @bar', 'params' => ['@bar' => 'bar'], 'formatted' => 'foo bar']],
```

Comment:
Please split this array into multiple lines (have a look at how other data providers are formatted). Also please always use the sprintf notation for the parameters, this way we can be sure that the function won't be called (if it does and the number of parameters is not equal or greater than the number of parameters used in the string it throws a warning)

Reply:
Not sure I'm following you, but by sprintf notation I'm guessing you mean e.g. `'foo %bar'` as the uninterpolated message (rather than `'foo @bar'`)?

Reply:
Exactly, since we're using `vsprintf` to format the message, using the `sprintf` notation we can be sure of these things:

- if the test case is testing that the message should be autoformatted but there are no parameters then the `vsprintf` function is not called
- if the test case is testing that the message should be autoformatted the `sprintf` function is called with the correct number of arguments
- if the test case is testing that the message should be equal to the `formatted` message passed from the user and there are parameters the `vsprintf` function is not called anyway and the correct message is used

Does it makes sense?